src/webgpu/api/validation/image_copy/buffer_related.spec.ts

```diff
@@ -0,0 +1,118 @@
+      .combine('bufferState', ['valid', 'destroyed', 'error'])
+  )
+  .fn(async t => {
+    const { method, bufferState } = t.params;
+
+    // A valid buffer.
+    let buffer = t.device.createBuffer({
+      size: 16,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+
+    switch (bufferState) {
+      case 'destroyed': {
+        buffer.destroy();
+        break;
+      }
+      case 'error': {
+        buffer = t.getErrorBuffer();
+        break;
+      }
+    }
+
+    const success = bufferState === 'valid';
+    const submit = bufferState === 'destroyed';
+
+    const texture = t.device.createTexture({
+      size: { width: 4, height: 4, depthOrArrayLayers: 1 },
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    t.testBuffer(
+      buffer,
+      texture,
+      { bytesPerRow: 0 },
+      { width: 0, height: 0, depthOrArrayLayers: 0 },
+      { dataSize: 16, method, success, submit }
+    );
+  });
+
+g.test('bytes_per_row_alignment')
+  .desc(
+    `Test that bytesPerRow must be a multiple of 256 for CopyB2T and CopyT2B if it is required.`
+  )
+  .params(u =>
+    u //
+      .combine('method', kImageCopyTypes)
+      .combine('dimension', ['2d', '3d'] as const)
+      .combine('format', kSizedTextureFormats)
+      .filter(({ dimension, format }) => textureDimensionAndFormatCompatible(dimension, format))
+      .filter(formatCopyableWithMethod)
+      .beginSubcases()
+      .combine('bytesPerRow', [undefined, 0, 1, 255, 256, 257, 512])
+      .combine('copyHeight', [0, 1, 2, 4, 8])
+      // copyHeight must be aligned with blockHeight
+      .filter(({ format, copyHeight }) => copyHeight % kTextureFormatInfo[format].blockHeight === 0)
+      // bytesPerRow must be specified and it must be equal or greater than the bytes size of each row if we are copying multiple rows.
+      // Note that we are copying one single block on each row in this test.
+      .filter(
+        ({ format, bytesPerRow, copyHeight }) =>
+          copyHeight / kTextureFormatInfo[format].blockHeight <= 1 ||
+          (bytesPerRow !== undefined && bytesPerRow >= kTextureFormatInfo[format].bytesPerBlock)
+      )
+  )
+  .fn(async t => {
+    const { method, dimension, format, bytesPerRow, copyHeight } = t.params;
+
+    const info = kTextureFormatInfo[format];
+    await t.selectDeviceOrSkipTestCase(info.feature);
+
+    const buffer = t.device.createBuffer({
+      size: 512 * 8 * 16,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+
+    let success = false;
+    // writeTexture doesn't require bytersPerRow to be 256-byte aligned.
+    if (method === 'WriteTexture') success = true;
+    // If the copy height <= 1, bytesPerRow is not required.
+    if (copyHeight / info.blockHeight <= 1 && bytesPerRow === undefined) success = true;
+    // If bytesPerRow > 0 and it is a multiple of 256, it will succeeed if other parameters are valid.
```

Comment:
If we are copying 0 or 1 row (copyHeight / info.blockHeight <= 1), we don't need to assign a value for rowsPerImage. But what happen if we assign a value for it? For example:
  1) bytesPerRow is 0 or a number that < bytes required for one row data for writeTexture (which doesn't require 256B aligned), is it valid? 
  2) bytesPerRow is 0 or a number that < bytes required for one row data and 256B aligned, it should be invalid, because the spec says "If specified, layout.bytesPerRow must be greater than or equal to bytesInLastRow." for B2T and T2B copies at 12.3.7 <Image Copies>.
 
  Considering the statement at 12.3.7 <Image Copies>, maybe we should add similar statement in order to clarify this edge case at 12.3.1 <GPUImageDataLayout>, WDYT?

Reply:
> * bytesPerRow is 0 or a number that < bytes required for one row data for writeTexture (which doesn't require 256B aligned), is it valid?

It should have the same rule as B2T and T2B; `writeTexture` also calls into the `validating linear texture data` algorithm - though when it fails, it throws an exception instead of a validation error, because the data is being read out of an ArrayBuffer instead of a GPUBuffer.

> Considering the statement at 12.3.7 , maybe we should add similar statement in order to clarify this edge case at 12.3.1 <GPUImageDataLayout>, WDYT?

Adding it here could be a useful reference, but it isn't strictly necessary. The validation rules need to be specified in algorithms anyway so it's clear what happens when they fail.